File: src/map.rs
use std::collections::HashMap;
use std::cmp::max;
use std::fmt;
use location::Location;
use room::Room;

pub type StringLiteral = &'static str;

// We want to be able to debug Room and `==` it.
#[derive(Debug, Eq, PartialEq)]
pub struct Map {
    pub title: String,
    pub rooms: HashMap<Location, Room>,
    max_x: u64, // east-most room
    max_y: u64, // north-most room
}

#[derive(Debug, Eq, PartialEq)]
pub struct ValidDirection {
    pub north: bool,
    pub south: bool,
    pub east: bool,
    pub west: bool,
}

impl Map {
    pub fn new(title: StringLiteral, room_list: Vec<Room>) -> Map {
        // Make a new mutable HashMap
        let mut rooms = HashMap::new();

        // Make new mutable max_x and max_y values, initialized to 0
        let mut max_x = 0;
        let mut max_y = 0;

        // Iterate over the room_list
        for room in room_list {
            // extract x and y from the room's location
            let x = room.location.x;
            let y = room.location.y;

            // update max_x and max_y if necessary
            max_x = max(max_x, x);
            max_y = max(max_y, y);

            // insert the room into the rooms Hash with its location
            // as a key. clone the location because removing the
            // location from the room to use as a key prevents use
            // from using the room as a value.
            rooms.insert(room.location.clone(), room);
        }

        // construct a new Map
        Map {
            title: title.to_string(),
            rooms: rooms,
            max_x: max_x,
            max_y: max_y,
        }
    }

    // valid_directions takes a Location contained in this
    // Map and answers which directions a player can go.
    // Specifically, a player is not allowed to move off
    // the edge of the map.
    pub fn valid_directions(&self, l: &Location) -> ValidDirection {
        ValidDirection {
            // identical to the Ruby code
            north: l.y < self.max_y,
            south: l.y > 0,
            east: l.x < self.max_x,
            west: l.x > 0,
        }
    }

    pub fn display_map(&self) {
        println!("Possible Destinations for {}", self.title);
        println!("{:=<1$}", "", self.title.len() + 26);
        for room_info in &self.rooms {
            println!("{}", room_info.1.name);
        }
    }
}

impl fmt::Display for Map {
    fn fmt(&self, _f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        // writeln!(_f, "{}", self.title)?;

        // for _ in self.title.chars() {
        //     write!(_f, "=")?;
        // }

        // write!(_f, "\n\n")?;

        // for row in &self.rooms {
        //     for room in row {
        //         write!(_f, "{:20} ", room.description)?;
        //     }

        //     write!(_f, "\n")?;

        //     for room in row {
        //         write!(_f, "{:20} ",
        //                format!("{}, {}", room.location.x, room.location.y))?;
        //     }

        //     write!(_f, "\n")?;
        // }

        Ok(())
    }
}

File: lib/map.rb
# A Map is an object that has a `title` as a string,
# a Hash of rooms (where the keys are Locations),
# a max_x, which is the east-most room, and max_y,
# which is the north-most room.
class Map < Struct.new(:title, :rooms, :max_x, :max_y)
  # A Map is initialized with a title and room list,
  # which is an array of rooms, and the room list is
  # converted into the expected Hash and max_x/max_y
  # values.
  def initialize(title, room_list, player)
    @player = player
    # Construct an empty Hash and initialize max_x and max_y
    rooms = {}
    max_x = 0
    max_y = 0

    # Iterate over the room_list
    room_list.each do |room|
      # extract x and y from the room's location
      x = room.location.x
      y = room.location.y

      #update max_x and max_y if necessary
      max_x = [max_x, x].max
      max_y = [max_y, y].max

      #insert the room into the rooms Hash with its location
      rooms[room.location] = room
    end

    # construct a new Map with title, rooms, max_x and max_y
    super title, rooms, max_x, max_y
  end

  def display_map
    puts title
    title.length.times do
      print "="
    end
    puts "\n"
    rooms.each_with_index do |room, index|
      if room[0] == @player.location
        puts "#{index}. #{room[1].name}. You are here."
      else
        puts "#{index}. #{room[1].name}"
      end
    end
  end

  def [](location)
    self.rooms[location]
  end

  # valid_directions takes a Location contained in this
  # Map and answers which directions a player can go.
  # Specifically, a player is not allowed to move off
  # the edge of the map.
  def valid_directions(location)
    {
      # a player can go north if they are not already
      # at the max_y
      north: location.y < self.max_y,
      # a player can go south if they are not at the
      # bottom edge
      south: location.y > 0,
      # a player can go east if they are not already
      # at the max_x
      east: location.x < self.max_x,
      # a player can go west if they are not at the
      # left edge
      west: location.x > 0
    }
  end
end